File: framec_tests/golang/var_scope/var_scope_test.go
package var_scope

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func (m *varScopeStruct) log(s string) {
	m.tape = append(m.tape, s)
}

func (m *varScopeStruct) do_nn() {
	m.Nn("|nn|[d]")
}
func (m *varScopeStruct) do_ny() {
	m.Ny("|ny|[d]")
}
func (m *varScopeStruct) do_yn() {
	m.Yn("|yn|[d]", "|yn|[x]")
}
func (m *varScopeStruct) do_yy() {
	m.Yy("|yy|[d]", "|yy|[x]")
}

func expected(state string, msg string, x string) []string {
	var result []string = []string{}

	result = append(result, "#.a")
	result = append(result, "$"+state+"[b]")
	result = append(result, "$"+state+".c")
	result = append(result, "|"+msg+"|"+"[d]")
	result = append(result, "|"+msg+"|"+".e")
	result = append(result, x)

	return result
}

func TestNoShadowing(t *testing.T) {
	sm := NewVarScope()
	x := sm.(*varScopeStruct)
	x.To_nn()
	x.do_nn()
	assert.Equal(t, expected("NN", "nn", "#.x"), x.tape)

}

func TestAllShadowingScenarios(t *testing.T) {
	sm := NewVarScope()
	x := sm.(*varScopeStruct)
	sm.To_nn()
	x.do_ny()
	assert.Equal(t, expected("NN", "ny", "|ny|.x"), x.tape)
	x.tape = x.tape[:0]
	x.do_yn()
	assert.Equal(t, expected("NN", "yn", "|yn|[x]"), x.tape)
	x.tape = x.tape[:0]
	x.do_yy()
	assert.Equal(t, expected("NN", "yy", "|yy|.x"), x.tape)

	sm = NewVarScope()
	x = sm.(*varScopeStruct)
	sm.To_ny()
	x.do_nn()
	assert.Equal(t, expected("NY", "nn", "$NY.x"), x.tape)
	x.tape = x.tape[:0]
	x.do_ny()
	assert.Equal(t, expected("NY", "ny", "|ny|.x"), x.tape)
	x.tape = x.tape[:0]
	x.do_yn()
	assert.Equal(t, expected("NY", "yn", "|yn|[x]"), x.tape)
	x.tape = x.tape[:0]
	x.do_yy()
	assert.Equal(t, expected("NY", "yy", "|yy|.x"), x.tape)

	sm = NewVarScope()
	x = sm.(*varScopeStruct)
	sm.To_yn()
	x.do_nn()
	assert.Equal(t, expected("YN", "nn", "$YN[x]"), x.tape)
	x.tape = x.tape[:0]
	x.do_ny()
	assert.Equal(t, expected("YN", "ny", "|ny|.x"), x.tape)
	x.tape = x.tape[:0]
	x.do_yn()
	assert.Equal(t, expected("YN", "yn", "|yn|[x]"), x.tape)
	x.tape = x.tape[:0]
	x.do_yy()
	assert.Equal(t, expected("YN", "yy", "|yy|.x"), x.tape)

	sm = NewVarScope()
	x = sm.(*varScopeStruct)
	sm.To_yy()
	x.do_nn()
	assert.Equal(t, expected("YY", "nn", "$YY.x"), x.tape)
	x.tape = x.tape[:0]
	x.do_ny()
	assert.Equal(t, expected("YY", "ny", "|ny|.x"), x.tape)
	x.tape = x.tape[:0]
	x.do_yn()
	assert.Equal(t, expected("YY", "yn", "|yn|[x]"), x.tape)
	x.tape = x.tape[:0]
	x.do_yy()
	assert.Equal(t, expected("YY", "yy", "|yy|.x"), x.tape)

}

func TestDisAmbiguation(t *testing.T) {}

File: framec_tests/cpp/var_scope/var_scope_test.cpp
#include <gtest/gtest.h>
#include <vector>
#include <string>
#include "var_scope.cpp"

class VarScopeController : public VarScope {
public:
    VarScopeController() : VarScope() {}

    void do_nn() {
        nn("|nn|[d]");
    }

    void do_ny() {
        ny("|ny|[d]");
    }

    void do_yn() {
        yn("|yn|[d]", "|yn|[x]");
    }

    void do_yy() {
        yy("|yy|[d]", "|yy|[x]");
    }
};

class VarScopeTest : public ::testing::Test {
protected:
    std::vector<std::string> expected(const std::string& state, const std::string& msg, const std::string& x) {
        std::vector<std::string> result;
        result.push_back("#.a");
        result.push_back("$" + state + "[b]");
        result.push_back("$" + state + ".c");
        result.push_back("|" + msg + "|" + "[d]");
        result.push_back("|" + msg + "|" + ".e");
        result.push_back(x);

        return result;
    }
};

TEST_F(VarScopeTest, TestNoShadowing) {
    VarScopeController sm;
    sm.to_nn();
    sm.do_nn();
    EXPECT_EQ(expected("NN", "nn", "#.x"), sm.tape);
}

TEST_F(VarScopeTest, TestAllShadowingScenarios) {
    VarScopeController sm;
    sm.to_nn();
    sm.do_ny();
    EXPECT_EQ(expected("NN", "ny", "|ny|.x"), sm.tape);

    sm.tape.clear();
    sm.do_yn();
    EXPECT_EQ(expected("NN", "yn", "|yn|[x]"), sm.tape);

    sm.tape.clear();
    sm.do_yy();
    EXPECT_EQ(expected("NN", "yy", "|yy|.x"), sm.tape);

    sm = VarScopeController();
    sm.to_ny();
    sm.do_nn();
    EXPECT_EQ(expected("NY", "nn", "$NY.x"), sm.tape);

    sm.tape.clear();
    sm.do_ny();
    EXPECT_EQ(expected("NY", "ny", "|ny|.x"), sm.tape);

    sm.tape.clear();
    sm.do_yn();
    EXPECT_EQ(expected("NY", "yn", "|yn|[x]"), sm.tape);

    sm.tape.clear();
    sm.do_yy();
    EXPECT_EQ(expected("NY", "yy", "|yy|.x"), sm.tape);

    sm = VarScopeController();
    sm.to_yn();
    sm.do_nn();
    EXPECT_EQ(expected("YN", "nn", "$YN[x]"), sm.tape);

    sm.tape.clear();
    sm.do_ny();
    EXPECT_EQ(expected("YN", "ny", "|ny|.x"), sm.tape);

    sm.tape.clear();
    sm.do_yn();
    EXPECT_EQ(expected("YN", "yn", "|yn|[x]"), sm.tape);

    sm.tape.clear();
    sm.do_yy();
    EXPECT_EQ(expected("YN", "yy", "|yy|.x"), sm.tape);

    sm = VarScopeController();
    sm.to_yy();
    sm.do_nn();
    EXPECT_EQ(expected("YY", "nn", "$YY.x"), sm.tape);

    sm.tape.clear();
    sm.do_ny();
    EXPECT_EQ(expected("YY", "ny", "|ny|.x"), sm.tape);

    sm.tape.clear();
    sm.do_yn();
    EXPECT_EQ(expected("YY", "yn", "|yn|[x]"), sm.tape);

    sm.tape.clear();
    sm.do_yy();
    EXPECT_EQ(expected("YY", "yy", "|yy|.x"), sm.tape);
}

int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

